### ISSUE TITLE
DJI Phantom 4 Pro Virtual Sticks Issues

### ISSUE THREAD

**Issue by charlieMonroe** [dji-sdk/Mobile-SDK-iOS#212]
We are using virtual sticks commands to control the drone and sometimes it stops reacting to any commands for yaw/roll/pitch, but vertical acceleration keeps working. Additionally, the drone does not respond to anything from the remote controller at this point - it generally seems frozen.

Additionally, if we start on the floor just by starting the motors (and not initiating an actual take-off), we are experiencing a weird behavior where we send positive value for vertical acceleration to the drone, trying to get it into the air. It seems to have a hard time doing so, however, and once it gets up, any other VS command causes the motors to suddenly stop and the drone to fall down. The flight controller state's altitude reports a negative altitude of -0.5 to -1.5 meters, while being about 10cm off the ground.

The ultrasonic sensors report the correct height of about 0.1 meter.

If needed, we can send additional information. When the app starts, we set the following:

- vision-assisted positioning on
- collision avoidance/obstakle avoidance off
- smart return to home off
- multiple flight mode off
- several camera-related settings
- virtual sticks on:
       - yaw control mode: angular velocity
       - roll/pitch control mode: velocity
       - roll/pitch coordinate system: body
       - VS advanced mode: on

After that Virtual Sticks commands are sent exactly as in your sample app every 0.05s (=20Hz).​

This is a huge issue that stongly undermines the reliability of your product. Please, let us know, what is the solution. Thank you!

-  DJI iOS SDK version: latest (4.3.2), but we've experienced these issues with 4.3 as well
-  DJI Product: Phantom 4 Pro, latest firmware from Dec 25 (1.04.0602)
-  iOS system version: 11.2.1
-  iOS device type: iPhone 6, iPad Pro 12.5'' (2017), iPad mini
-  Xcode version: Xcode 9.2

**Comment by PandaraWen** [comment]
You can send email to dji team to get response much quickly

**Comment by charlieMonroe** [comment]
Done that as well, but someone might have run into the same issues in the past...

**Comment by oliverou** [comment]
Hi @charlieMonroe , we have just released the iOS SDK 4.4, could you help to check if this issue still exist?

**Comment by charlieMonroe** [comment]
@oliverou - sorry, we did try 4.4 as well, same issues. This more or less seems to be a firmware issue to us as the `DJIVirtualStickFlightControlData` structure is successfully sent (the callback is invoked with no error), but the aircraft will not move accordingly. Additionally, the remote controller is then not reacting at all. When the virtual sticks are disabled (the same setup, the same app, just disabling the VS), flying the Phantom using remote controller works just fine - is a different communication approach used here?

**Comment by oliverou** [comment]
Hi @charlieMonroe, I help to updated the latest firmware and tested it with the latest iOS SDK 4.4. I use the joystick function and found the drones could move according to the command we send. I didn't find the problem of "Frozen" as you mentioned. 

Besides, if you found the joystick doesn't work, you could still control the drone through the remote controller by switching the mode back and forth to the **P** mode to regain the control access.

Hope this help!

**Comment by aksh1t** [comment]
@charlieMonroe Have you seen this behavior replicate on multiple drones? Knowing that this issue happens on just one or multiple drones might help pinpoint if this is an SDK thing or something wrong with hardware on one drone.

**Comment by XietongLU** [comment]
@charlieMonroe please check if you turned on the obstacle avoidance. If so, turn it off and retry. If the problem still exists, I think you need to send us the flight log from the aircraft (through Assistant 2).

**Comment by charlieMonroe** [comment]
Turning off obstacle avoidance is one of the first things that are set when the app launches. Logs sent to you via email conversation.

**Comment by gabejc** [comment]
We have seen similar behavior. We will be issuing virtual stick velocity commands, and then suddenly the drone stops responding to them, but the callback reports no error. We have not been able to identify how to reliably reproduce this situation...it seems to happen on perhaps ~2% of our flights.

This has happened on every version of iOS, XCode, and the DJI SDK we have used, dating back more than 1.5 years ago. We've seen it with Phantom4s, Inspire 1s, and M200s.

Most recent occurrence
* Drone: Phantom4 Pro
* DJI SDK: 4.4
* XCode: 9.2
* Device: iPad Pro 9.7"
* iOS: 11.2.2

I agree with @charlieMonroe that this is a huge issue, and we are very interested in its resolution.

**Comment by chaitanyantr** [comment]
i have been working on dji sdk 3.5 for an year.Virtual stick has been worked all the times when(sample app reference).Am i missing any sequence of enable virtual stick,any new release on that?

**Comment by geetesh** [comment]
In my case, its definitely the obstacle avoidance making the aircraft brake. 
My case:
Aircraft: Mavic 2 zoom / Spark
Device: Ipad iOS 12.2
I checked via the VisionControlState as mentioned in the API and it shows when it is actively braking the aircraft. You can check via RC sticks too. If you move the aircraft the brake=0 unless there is an obstacle around. If there is an obstacle brake = 1 but of obstacle avoidance is turned off it follows the stick commands. 
However, if you are in virtual joystick mode, even after disabling the obstacle avoidance, the aircraft brakes. Every now and then when obstacle is removed the brake=0 and aircraft starts following the virtual joystick command. 
So, what is the real way to disable obstacle avoidance in virtual joystick mode? If anyone has idea, please suggest.

**Comment by dmtburke** [comment]
Also having this problem on android

**Comment by zoemcmac** [comment]
method setActiveObstacleAvoidanceEnabled:withCompletion
https://developer.dji.com/api-reference/ios-api/Components/IntelligentFlightAssistant/DJIIntelligentFlightAssistant.html?search=obstacle&i=6&#djiintelligentflightassistant_setactiveobstacleavoidanceenabled_inline

turn that off first?

**Comment by geetesh** [comment]
One user confirms it working on android though!
https://stackoverflow.com/questions/55689833/dji-mobile-sdk-how-to-disable-obstacle-avoidance-in-virtual-joystick-control-mo

**Comment by dmtburke** [comment]
I don't think that it's collision detection causing problems in my case. I've had the drone fly into things unintentionally before, but I can verify that again the next time I test. My application is intended to fly the drone and do certain tasks without user involvement, keeping track of where it goes (testing in a garage, gps signal is not reliable) by taking the speed that is set while sending virtual stick tasks. Today I ran a huge number of tests with the drone running the same set of tasks over and over again (nearby obstacles were constant). It ran them fine most of the time, but every once in a while it would be a problem.

When the virtual sticks stop working (this is somewhat rare, but still dangerous) and don't report an error, the system thinks that everything is going fine, and when it tries to go somewhere else its position in real life, and its position in the application are no longer synced, meaning the drone will likely fly into things. I have made some changes to these position updates so that they're independent of what is being sent to the virtual sticks (polling the FlightControllerState's velocity), making it at least safe to fly. However, detecting that a problem occurred is turning into a lot of work, considering that the callbacks should be saying something.

I'm using android and a mavic air. I should clarify that I'm not looking for support here, just thought I'd mention that the virtual sticks callbacks also have a problem in the android version, maybe it will help devs isolate the problem. I'll probably file this as an android issue soon, but I'm done working on it for the day.

**Comment by LGstudio** [comment]
I have a similar issues on android with sdk 4.8.1 and Mavic 2 Zoom
I would like to send commands via virtual sticks, but also enable the user to make correction.
When I set ..virtualSticksEnabled(true...), after a first automatic command, the remote controller sticks work no more. I was trying the controller state callback, but the joystick states are constantly remain 0, after the first automatic movement. I would also like to lower the joystick sensitivity, but in enabled tripod mode automatic commands do not work at all
I am a big trouble finding a working and reliable solution.

**Comment by aeriu** [comment]
Do you have any update on this? I have similar issues with Mavic 2 Pro and DJI SDK 4.10. Is there any beta release with bugfix?

**Comment by zoemcmac** [comment]
Virtual Sticks are a special kind of mission, that turn off the joysticks completely. (Normally the operator could accelerate through a way point mission with a forward stick command.)
Like all missions, manually toggling the flight mode switch breaks out of mission mode so a pilot can regain control.
I would suggest your UX could throw the user a pop up tutorial about this feature.  Its good for pilots to know that even a return to home can be broken off by toggling that switch. 
The user will not be able to resume the mission, but at least they can avoid the tree or water they didn’t realize could be an issue.

**Comment by mclui** [comment]
I'm seeing this very same issue using the Onboard SDK.

> Additionally, if we start on the floor just by starting the motors (and not initiating an actual take-off), we are experiencing a weird behavior where we send positive value for vertical acceleration to the drone, trying to get it into the air. It seems to have a hard time doing so, however, and once it gets up, any other VS command causes the motors to suddenly stop and the drone to fall down. 

Has there been any more insight about the cause?
